translator/internal/internal_to_oc.go

```diff
@@ -0,0 +1,432 @@
+	var ocKind tracetranslator.OpenTracingSpanKind
+	switch kind {
+	case data.SpanKindCONSUMER:
+		ocKind = tracetranslator.OpenTracingSpanKindConsumer
+	case data.SpanKindPRODUCER:
+		ocKind = tracetranslator.OpenTracingSpanKindProducer
+	case data.SpanKindUNSPECIFIED:
+	case data.SpanKindINTERNAL:
+	case data.SpanKindSERVER: // explicitly handled as SpanKind
+	case data.SpanKindCLIENT: // explicitly handled as SpanKind
+	default:
+
+	}
+
+	if string(ocKind) == "" {
+		// No matching kind attribute value
+		return nil
+	}
+
+	return stringAttributeValue(string(ocKind))
+}
+
+func stringAttributeValue(val string) *octrace.AttributeValue {
+	return &octrace.AttributeValue{
+		Value: &octrace.AttributeValue_StringValue{
+			StringValue: &octrace.TruncatableString{
+				Value: val,
+			},
+		},
+	}
+}
+
+// OTLP follows the W3C format, e.g. "vendorname1=opaqueValue1,vendorname2=opaqueValue2"
+func traceStateToOC(traceState data.TraceState) *octrace.Span_Tracestate {
+	if traceState == "" {
+		return nil
+	}
+
+	// key-value pairs in the "key1=value1" format
+	pairs := strings.Split(string(traceState), ",")
+
+	entries := make([]*octrace.Span_Tracestate_Entry, 0, len(pairs))
+	for _, pair := range pairs {
+		kv := strings.SplitN(pair, "=", 2)
+		if len(kv) == 0 {
+			continue
+		}
+
+		key := kv[0]
+		val := ""
+		if len(kv) >= 2 {
+			val = kv[1]
+		}
+
+		entries = append(entries, &octrace.Span_Tracestate_Entry{
+			Key:   key,
+			Value: val,
+		})
+	}
+
+	return &octrace.Span_Tracestate{
+		Entries: entries,
+	}
+}
+
+func spanKindToOC(kind data.SpanKind) octrace.Span_SpanKind {
+	switch kind {
+	case data.SpanKindSERVER:
+		return octrace.Span_SERVER
+	case data.SpanKindCLIENT:
+		return octrace.Span_CLIENT
+	// NOTE: see `spanKindToOCAttribute` function for custom kinds
+	case data.SpanKindUNSPECIFIED:
+	case data.SpanKindINTERNAL:
+	case data.SpanKindPRODUCER:
+	case data.SpanKindCONSUMER:
+	default:
+	}
+
+	return octrace.Span_SPAN_KIND_UNSPECIFIED
+}
+
+func eventsToOC(events []*data.SpanEvent, droppedCount uint32) *octrace.Span_TimeEvents {
+	if len(events) == 0 && droppedCount == 0 {
+		return nil
+	}
+
+	ocEvents := make([]*octrace.Span_TimeEvent, 0, len(events))
+	for _, event := range events {
+		ocEvents = append(ocEvents, eventToOC(event))
+	}
+
+	return &octrace.Span_TimeEvents{
+		TimeEvent:                 ocEvents,
+		DroppedMessageEventsCount: int32(droppedCount),
+	}
+}
+
+func eventToOC(event *data.SpanEvent) *octrace.Span_TimeEvent {
+	attrs := event.Attributes()
+
+	// Consider TimeEvent to be of MessageEvent type if all and only relevant attributes are set
+	ocMessageEventAttrs := []string{
+		conventions.OCTimeEventMessageEventType,
+		conventions.OCTimeEventMessageEventID,
+		conventions.OCTimeEventMessageEventUSize,
+		conventions.OCTimeEventMessageEventCSize,
+	}
+	if len(attrs) == len(ocMessageEventAttrs) {
+		ocMessageEventAttrValues := map[string]data.AttributeValue{}
```

Comment:
since the keys come in a fixed order you could do with an array instead, but probably not worth the trouble (better do some benchmark before changing it)